src/lib.rs

```diff
@@ -0,0 +1,171 @@
+extern crate regex;
+
+use regex::Regex;
+
+#[macro_use]
+use std::fs::File;
+use std::io::prelude::*;
+use std::collections::HashMap;
+
+#[derive(Debug, PartialEq)]
+pub enum IdentType {
+    Space,
+    Tab,
+    None
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Indent {
+    amount: usize,
+    indent: String,
+    ident_type: IdentType
+}
+
+pub fn detect_indent(string: &str) -> Indent {
+    // TODO: actually implement this
+    let mut weight = 0;
+    let mut amount = 0; // TODO
+    let indent_regex = Regex::new(r"^(?:( )+|\t+)").unwrap();
+    // let mut indents = HashMap::new();
+
+    // indents.entry(1).or_insert(Vec::new()).push(2);
+    // indents.entry(1).or_insert(Vec::new()).push(3);
+
+    let mut prev = 0;
+    for line in string.lines() {
+        amount = 4;
+        if line.is_empty() { continue; }
+        let mut depth = 0;
+
+        match indent_regex.captures(line) {
+            Some(captures) => {
+                depth = captures.len();
+
+                captures.get(0).map( |capture| {
+                    match capture.as_str() {
+                        " "  => weight +=1,
+                        "\n" => weight -=1,
+                        _ => ()
+                    }
+                });
+
+                // indents.entry(indent).push(line);
+                // how much did it increase or decreae, that is the current indentation. Store it
+                // in the map
+            }
+
+            None => depth = 0
+        }
+
+        prev = depth;
+    }
+
+    let ident_type;
+    let indent;
+
+    if amount == 0 {
+        ident_type = IdentType::None;
+        indent = "".to_string();
+    } else {
+        if weight >= 0 {
+            ident_type = IdentType::Space;
+            indent = " ".repeat(amount);
+        } else {
+            ident_type = IdentType::Tab;
+            indent = "\t".repeat(amount);
+        }
+    }
+
+    // println!("type: {:?} amount: {:?}", ident_type, amount);
+
+    return Indent {
+        amount,
+        indent,
+        ident_type
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // TODO: better panics/errors
+    fn string_from_file(filepath: &str) -> String {
+        let mut contents = String::new();
+        match File::open(filepath) {
+            Ok(mut file) => {
+                match file.read_to_string(&mut contents) {
```

Comment:
I like to use the [readext](https://crates.io/crates/readext) crate so I can just say e.g. `let contents = file.read_into_string().expect("no contents");` instead of having to create the string on one line and then mutate it. I hope that API eventually makes it into std.